### PR TITLE
feat: allow plugins to specify if they should be transformed

### PR DESCRIPTION
**Pull request by sheremet-va**
Related: https://github.com/vitest-dev/vitest/issues/369, https://github.com/vitest-dev/vitest/issues/264

I guess plugins can use `config` hook to override this, right?

### CODE REVIEW

File: packages/vitest/src/node/transform.ts [hunk 1]
```diff
@@ -21,7 +21,8 @@ export async function transformRequest(ctx: Vitest, id: string) {
 async function _transformRequest(ctx: Vitest, id: string) {
   let result: TransformResult | null = null
 
-  if (id.match(/\.(?:[cm]?[jt]sx?|json)$/)) {
+  // Vue plugin can generate this url: "?vue&type=template&lang.js", so wee need to also check for ?
```
Comment (sheremet-va) [inline packages/vitest/src/node/transform.ts:24]: or Vue plugin can declare it `ssrTransformInclude` as `/\?vue/`

File: examples/vue-jsx/vite.config.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import { defineConfig } from 'vite'
+import Vue from '@vitejs/plugin-vue'
+import Jsx from '@vitejs/plugin-vue-jsx'
+
+export default defineConfig({
+  plugins: [Vue(), Jsx()],
+  test: {
+    global: true,
+    environment: 'happy-dom',
+    ssrTransformInclude: [/.[tj]sx$/],
```
Comment (antfu) [inline examples/vue-jsx/vite.config.ts:10]: Have one side include isn't enough, we could have:

```ts
{
  transformMode: {
    ssr: [...], // directly ssr transform
    client: [...], // client transform + ssr transform
  }
}
```

Comment (sheremet-va) [reply examples/vue-jsx/vite.config.ts:10]: I am a little bit confused with terminology here

ssr is this? https://github.com/vitest-dev/vitest/pull/469/files#diff-cdd740fae25ab6905acf643dbfb4336f419a6fdfe4a6c73e61868063e1cc5abeR25

and client is this? https://github.com/vitest-dev/vitest/pull/469/files#diff-cdd740fae25ab6905acf643dbfb4336f419a6fdfe4a6c73e61868063e1cc5abeL28-L32

Comment (sheremet-va) [reply examples/vue-jsx/vite.config.ts:10]: Also one contradicts another, so do wee really need this? Or you have something else in mind concerning how we transform files?

Comment (sheremet-va) [reply examples/vue-jsx/vite.config.ts:10]: Or you mean something like this?

```ts
if(
  (id.match(...) /* current matcher */ || isIncludes(config.transformMode.client, id)) 
  && !isIncludes(config.transformMode.ssr, id)
) {
  transformRequest()
} else {
  transformRequest()
  ssrTransform()
}
```

Comment (antfu) [reply examples/vue-jsx/vite.config.ts:10]: Let me do it

Comment (sheremet-va) [reply examples/vue-jsx/vite.config.ts:10]: I am letting you do it! But I would still like an explanation 👀

File: packages/vitest/src/node/transform.ts [hunk 3]
```diff
@@ -18,28 +18,40 @@ export async function transformRequest(ctx: Vitest, id: string) {
   return promiseMap.get(id)
 }
 
+function getTransformMode(ctx: Vitest, id: string) {
+  const withoutQuery = id.split('?')[0]
```
Comment (sheremet-va) [inline packages/vitest/src/node/transform.ts:22]: why without query?

File: packages/vitest/src/node/transform.ts [hunk 4]
```diff
@@ -18,28 +18,40 @@ export async function transformRequest(ctx: Vitest, id: string) {
...
+
+  if (ctx.config.transformMode?.web?.some(r => withoutQuery.match(r)))
```
Comment (sheremet-va) [inline packages/vitest/src/node/transform.ts:24]: or rather why this withour query

Comment (sheremet-va) [reply packages/vitest/src/node/transform.ts:24]: as a plugin author I would expect it to have full path 👀

File: packages/vitest/src/node/transform.ts (hunk 3 above)
Comment (antfu) [reply packages/vitest/src/node/transform.ts:22]: Why with query then?

I think the original approach is a bit problematic as I didn't consider the query, as `.vue` should be treated as client transform but `.vue?vue&lang.ts` be treated as ssr transform is weird.

Also, a query could be like `.ts?v=xxx` that we should generally not consider them I think

Comment (sheremet-va) [reply packages/vitest/src/node/transform.ts:22]: can it be a `?v=` tho? I think by default we should check withour query, but give authors flexiblity? 
but maybe I'm thinking too much into it.

Comment (sheremet-va) [reply packages/vitest/src/node/transform.ts:22]: can't really seem to find a proper usecase where it would be needed to have a query

**Pull request merged**